src/java.base/share/classes/sun/security/util/DerValue.java

```diff
@@ -342,87 +424,23 @@ public DerValue(InputStream in) throws IOException {
-        case tag_T61String:
-            charset = ISO_8859_1;
-            break;
-        case tag_BMPString:
-            charset = UTF_16BE;
-            break;
-        case tag_UTF8String:
-            charset = UTF_8;
-            break;
-        case tag_UniversalString:
-            charset = Charset.forName("UTF_32BE");
-            break;
-        default:
-            throw new IllegalArgumentException("Unsupported DER string type");
-        }
-
-        byte[] buf = value.getBytes(charset);
-        length = buf.length;
-        buffer = new DerInputBuffer(buf, true);
-        DerInputStream result = new DerInputStream(buffer);
-        result.mark(Integer.MAX_VALUE);
-        return result;
-    }
-
-    /*
-     * helper routine
+    /**
+     * Encode an ASN1/DER encoded datum onto a DER output stream.
      */
-    private DerInputStream init(boolean fullyBuffered, InputStream in,
-        boolean allowBER) throws IOException {
-
-        tag = (byte)in.read();
-        byte lenByte = (byte)in.read();
-        length = DerInputStream.getLength(lenByte, in);
-        if (length == -1) { // indefinite length encoding found
-            in = new ByteArrayInputStream(
-                    DerIndefLenConverter.convertStream(in, lenByte, tag));
-            if (tag != in.read())
-                throw new IOException
-                        ("Indefinite length encoding not supported");
-            length = DerInputStream.getDefiniteLength(in);
-        }
-
-        if (fullyBuffered && in.available() != length)
-            throw new IOException("extra data given to DerValue constructor");
-
-        byte[] bytes = IOUtils.readExactlyNBytes(in, length);
-
-        buffer = new DerInputBuffer(bytes, allowBER);
-        return new DerInputStream(buffer);
+    public void encode(DerOutputStream out) throws IOException {
+        out.write(tag);
+        out.putLength(end - start);
+        out.write(buffer, start, end - start);
+        data.pos = data.end; // Compatibility. Reach end.
     }
 
     /**
-     * Encode an ASN1/DER encoded datum onto a DER output stream.
+     * Returns a new DerInputStream pointing at the start of buffer.
+     *
+     * @return the new DerInputStream value
      */
-    public void encode(DerOutputStream out)
-    throws IOException {
-        out.write(tag);
-        out.putLength(length);
-        // XXX yeech, excess copies ... DerInputBuffer.write(OutStream)
-        if (length > 0) {
-            byte[] value = new byte[length];
-            // always synchronized on data
-            synchronized (data) {
-                buffer.reset();
-                if (buffer.read(value) != length) {
-                    throw new IOException("short DER value read (encode)");
-                }
-                out.write(value);
-            }
-        }
+    public final DerInputStream data() {
+        return new DerInputStream(this);
     }
 
     public final DerInputStream getData() {
```

Comment:
There are a lot of classes calling this method and it's different from `data()`. `data()` returns a new stream always pointing to the 1st sub-value in the content, but `getData()` returns the data field which could advance after a read call. This means you can call `getData().getDerValue()` to read the 1st sub-value, and then `getData().getDerValue()` again to read the 2nd. Quite unobvious but the actual behavior. I can add some comments. I named the new method `data()` because I want the future nice-behavior method looks simple.

Reply:
Both the data field and `getData()`are public and seems interchangeable (redundant?). With this new method named data(), it sounds like a replacement for the data field which I am not sure if this is your intention or not. Do we really needs data()? There is the subs() method for returning the sub values, no?

Reply:
I suggest future app uses data() only. subs() might not be easy to use (Ex: when dealing with OPTIONAL fields), otherwise DerInputStream will be useless.